File: src/Scripting/Ruby/patterns/leader_followers.rb
# frozen_string_literal: true

module RubyPatterns
  module LeaderFollowers
    module_function

    def run
      events = %w[one two]
      workers = %i[leader follower].each
      handled = events.map { |event| [workers.next, event] }
      expected = [[:leader, 'one'], [:follower, 'two']]
      raise 'Leader / Followers failed' unless handled == expected
      true
    end
  end
end

RubyPatterns::LeaderFollowers.run if $PROGRAM_NAME == __FILE__

File: src/Scripting/PHP/patterns/leader_followers.php
<?php declare(strict_types=1);
$check = static fn(bool $ok) => $ok ?: throw new RuntimeException('leader_followers');
$events = new SplQueue(); foreach (['a','b','c'] as $e) $events->enqueue($e);
$workers = ['w1','w2']; $handled = [];
$i = 0; while (!$events->isEmpty()) { $leader = $workers[$i++ % 2]; $handled[] = $leader.':'.$events->dequeue(); }
$check($handled === ['w1:a','w2:b','w1:c']);
